pandas/tests/reshape/concat/test_index.py

```diff
@@ -265,3 +265,33 @@ def test_concat_multiindex_dfs_with_deepcopy(self):
         tm.assert_frame_equal(result_copy, expected)
         result_no_copy = concat(example_dict, names=["testname"])
         tm.assert_frame_equal(result_no_copy, expected)
+
+    @pytest.mark.parametrize(
+        "mi1",
+        [
+            MultiIndex.from_product([["a"], range(2)]),
+            MultiIndex.from_product([["a"], np.arange(2.0)]),
+            MultiIndex.from_product([["b"], ["A", "B"]]),
+            MultiIndex.from_product(
+                [["c"], pd.date_range(start="2017", end="2018", periods=2)]
+            ),
+        ],
+    )
+    @pytest.mark.parametrize(
+        "mi2",
+        [
+            MultiIndex.from_product([["a"], range(2)]),
+            MultiIndex.from_product([["a"], np.arange(2.0)]),
+            MultiIndex.from_product([["b"], ["A", "B"]]),
+            MultiIndex.from_product(
+                [["c"], pd.date_range(start="2017", end="2018", periods=2)]
+            ),
+        ],
+    )
+    def test_concat_with_various_multiindex_dtypes(self, mi1, mi2):
+        # GitHub #23478
+        df1 = DataFrame(np.zeros((1, len(mi1))), columns=mi1)
+        df2 = DataFrame(np.zeros((1, len(mi2))), columns=mi2)
+
+        with tm.assert_produces_warning(None):
+            concat((df1, df2), axis=1)
```

Comment:
can we check this against the expectations (might be slightly tricky with all of the paramaterizations but otherwise hard to know if the results are actually right)..

Reply:
I have added the expected result using ```MultiIndex.from_tuples```. However, I am getting the following issue
```python
>>> mi1
MultiIndex([('a', 0),
            ('a', 1)],
           )
>>> mi2 
MultiIndex([('a', 0.0),
            ('a', 1.0)],
           )
>>> df1 = DataFrame(np.zeros((1, len(mi1))), columns=mi1)
>>> df1
     a     
     0    1
0  0.0  0.0
>>> df2 = DataFrame(np.zeros((1, len(mi2))), columns=mi2)
>>> df2
     a
   0.0  1.0
0  0.0  0.0
>>>expected = DataFrame(np.zeros((1, len(mi1) + len(mi2))), columns=MultiIndex.from_tuples(list(mi1) + list(mi2)))
>>>expected
     a
     0    1    0    1
0  0.0  0.0  0.0  0.0
>>>result = concat((df1, df2), axis=1)
>>>result
     a
   0.0  1.0  0.0  1.0
0  0.0  0.0  0.0  0.0
```
Using ```MultiIndex.from_tuples``` converts the datatypes of all columns to integers. Using ```concat``` converts the datatypes of all columns to floats so ```tm.assert_frame_equal(expected, result)``` fails. Is this a bug? should the type conversions be more consistent?